src/dispatch/plugins/dispatch_slack/middleware.py

```diff
@@ -0,0 +1,221 @@
+    """Attempts to determine the current context of the event."""
+    context.update({"subject": SubjectMetadata(channel_id=context["channel_id"])})
+    await next()
+
+
+async def button_context_middleware(payload, context, next):
+    """Attempt to determine the current context of the event."""
+    context.update({"subject": SubjectMetadata.parse_raw(payload["value"])})
+    await next()
+
+
+async def action_context_middleware(body, context, next):
+    """Attempt to determine the current context of the event."""
+    context.update({"subject": SubjectMetadata.parse_raw(body["view"]["private_metadata"])})
+    await next()
+
+
+async def message_context_middleware(context, next):
+    """Attemps to determine the current context of the event."""
+    conversation = resolve_conversation_from_context(context["channel_id"])
+    if conversation:
+        context.update(
+            {
+                "subject": SubjectMetadata(
+                    type="incident",
+                    id=conversation.incident.id,
+                    organization_slug=conversation.incident.project.organization.slug,
+                    project_id=conversation.incident.project.id,
+                ),
+                "db_session": Session.object_session(conversation),
+            }
+        )
+    else:
+        raise ContextError("Unable to determine context for message.")
+
+    await next()
+
+
+async def restricted_command_middleware(context, db_session, user, next):
+    """Rejects commands from unauthorized individuals."""
+    allowed_roles = [ParticipantRoleType.incident_commander, ParticipantRoleType.scribe]
+    participant = participant_service.get_by_incident_id_and_email(
+        db_session=db_session, incident_id=context["subject"].id, email=user.email
+    )
+
+    # if any required role is active, allow command
+    for active_role in participant.active_roles:
+        for allowed_role in allowed_roles:
+            if active_role.role == allowed_role:
+                return await next()
+
+    raise RoleError(
+        f"User does not have correct role. allowedRoles: {','.join(['r.name for r in allowed_roles'])}"
+    )
+
+
+async def user_middleware(body, payload, db_session, client, context, next):
+    """Attempts to determine the user making the request."""
+
+    # for modals
+    if body.get("user"):
+        user_id = body["user"]["id"]
+
+    # for messages
+    if payload.get("user"):
+        user_id = payload["user"]
+
+    # for commands
+    if payload.get("user_id"):
+        user_id = payload["user_id"]
+
+    if not user_id:
+        raise ContextError("Unabled to determine user from context.")
+
+    email = (await client.users_info(user=user_id))["user"]["profile"]["email"]
+    context["user"] = user_service.get_or_create(
+        db_session=db_session,
+        organization=context["subject"].organization_slug,
+        user_in=DispatchUser(email=email),
+    )
```

Comment:
It looks like an unregistered user interacting with Dispatch Slack application may raise an exception here. Further down in the `create()` function,`user_in.password` will be set to `None`:

https://github.com/Netflix/dispatch/blob/1bd965dcf636080793792cbede00b79a00ef6389/src/dispatch/auth/service.py#L148 -> Raises: `TypeError: encoding without a string argument`

I encountered this when I attempted to report an incident using the `/dispatch-report-incident` slash command from a ~personal email that wasn't registered in Dispatch. Not sure of the expected behavior here, but figured I would call it out.

Reply:
The password attribute should always be set by this validator:
https://github.com/Netflix/dispatch/blob/1bd965dcf636080793792cbede00b79a00ef6389/src/dispatch/auth/models.py#L145

Did you have an exception you could point me to?

Reply:
```bash
ERROR:encoding without a string argument:/Users/wshel/Projects/dispatch/src/dispatch/plugins/dispatch_slack/bolt.py:errors:43
Traceback (most recent call last):
  File "/Users/wshel/.pyenv/versions/3.9.7/envs/dispatch/lib/python3.9/site-packages/slack_bolt/app/async_app.py", line 532, in async_dispatch
    ) = await listener.run_async_middleware(req=req, resp=resp)
  File "/Users/wshel/.pyenv/versions/3.9.7/envs/dispatch/lib/python3.9/site-packages/slack_bolt/listener/async_listener.py", line 53, in run_async_middleware
    resp = await m.async_process(req=req, resp=resp, next=_next)
  File "/Users/wshel/.pyenv/versions/3.9.7/envs/dispatch/lib/python3.9/site-packages/slack_bolt/middleware/async_custom_middleware.py", line 45, in async_process
    return await self.func(
  File "/Users/wshel/Projects/dispatch/src/dispatch/plugins/dispatch_slack/middleware.py", line 122, in user_middleware
    context["user"] = user_service.get_or_create(
  File "/Users/wshel/Projects/dispatch/src/dispatch/auth/service.py", line 181, in get_or_create
    user = create(db_session=db_session, organization=organization, user_in=user_in)
  File "/Users/wshel/Projects/dispatch/src/dispatch/auth/service.py", line 148, in create
    password = bytes(user_in.password, "utf-8")
TypeError: encoding without a string argument
DEBUG:encoding without a string argument:/Users/wshel/Projects/dispatch/src/dispatch/plugins/dispatch_slack/bolt.py:errors:44
```

Reply:
Ah got it. I'm using the wrong model. I should use `DispatchRegister`, not `DispatchUser`. I can make that change.

https://github.com/Netflix/dispatch/blob/1bd965dcf636080793792cbede00b79a00ef6389/src/dispatch/auth/service.py#L175